packages/@lwc/engine-core/src/framework/template.ts

```diff
@@ -113,6 +115,38 @@ function validateLightDomTemplate(template: Template, vm: VM) {
     }
 }
 
+export function parseFragment(strings: string[], ...keys: number[]): () => Element {
+    return function (): Element {
+        const {
+            context: { hasScopedStyles, stylesheetToken },
+            shadowMode,
+        } = getVMBeingRendered()!;
+        const classToken = hasScopedStyles && stylesheetToken ? ' ' + stylesheetToken : '';
+        const classAttrToken =
+            hasScopedStyles && stylesheetToken ? ` class="${stylesheetToken}"` : '';
+        const attrToken =
+            stylesheetToken && shadowMode === ShadowMode.Synthetic ? ' ' + stylesheetToken : '';
+
+        const htmlFragments: string[] = [];
+        for (let i = 0, n = keys.length; i < n; i++) {
+            switch (keys[i]) {
+                case 0: // styleToken in existing class attr
+                    htmlFragments.push(strings[i], classToken);
+                    break;
+                case 1: // styleToken for added class attr
+                    htmlFragments.push(strings[i], classAttrToken);
+                    break;
+                case 2: // styleToken as attr
+                    htmlFragments.push(strings[i], attrToken);
+            }
+        }
+
+        htmlFragments.push(strings[strings.length - 1]);
```

Comment:
```suggestion
        let htmlFragment = '';
        for (let i = 0, n = keys.length; i < n; i++) {
            switch (keys[i]) {
                case 0: // styleToken in existing class attr
                    htmlFragment += strings[i] + classToken;
                    break;
                case 1: // styleToken for added class attr
                    htmlFragment += strings[i] + classAttrToken;
                    break;
                case 2: // styleToken as attr
                    htmlFragment += strings[i] + attrToken;
            }
        }

        htmlFragment += strings[strings.length - 1];

        return createFragment(htmlFragment);
```

It should be slightly faster to use string concatenation rather than an array join, although I admit that when I tested in Chrome+Tachometer I didn't see a >=1% difference.